README.md
fake-secretsmanager
===================

`fake-secretsmanager` is a stand-in for the full [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/) for testing, local development, kitchen runs, and other such things where using the real deal is neither needed nor desirable.

It should go without saying that this absolutely should not be used for production (or probably even staging) - it does not use SSL to encrypt traffic, and its "secrets" storage is in a plain JSON file.

Installation
------------

To build `fake-secretsmanager` from source, assuming you have Go installed and your Go development environment set up properly, run:

```
$ go get -u github.com/apexlearning/fake-secretsmanager
```

`fake-secretsmanager` has been built using golang 1.10. It may build with earlier or later versions of the compiler, but this has not been tested.

A `Dockerfile` is also provided, if running `fake-secretsmanager` inside docker is more convenient. Build it with `docker build -t your-handy-name/fake-secretsmanager .`, and run it with:

```
$ docker run -p 7887:7887 -v /path/to/your/secrets.json:/opt/fake-secretsmanager/data/secrets.json --name fakesm -d your-handy-name/fake-secretsmanager
```

Usage
-----

```
Usage:
  fake-secretsmanager [OPTIONS]

Application Options:
  -v, --version       Print version info.
  -a, --addr=         IP address to listen on. Default: ':7887'. [$FAKESM_ADDR]
  -f, --secrets-json= Path to JSON file containing the secrets in a hash. The
                      JSON hash key names are the secret names. If the secret
                      is itself JSON, it needs to be escaped and stuffed in
                      there as a normal string. [$FAKESM_SECRETS_JSON]

Help Options:
  -h, --help          Show this help message
```

The `secrets.json` file provides an example of how to format the JSON file that stores the secrets. Each secret must be a string. If it is not a string, it must be properly quoted and escaped as per the example escaped JSON in that file.

To use `fake-secretsmanager`, supply a custom endpoint to the AWS cli or in your code's secretsmanager client constructor.

CLI:

```
$ aws secretsmanager get-secret-value --endpoint-url http://localhost:7887 --secret-id foo/json/escaped/sssssh

$ aws secretsmanager list-secrets --endpoint http://localhost:7887
```

Ruby:

```
secretsmanager = Aws::SecretsManager::Client.new(
  region: my_region,
  endpoint: 'http://localhost:7887',
  # .....
)
```

Supported Functionality
-----------------------

Currently, `fake-secretsmanager` supports the following [AWS Secrets Manager API](https://docs.aws.amazon.com/secretsmanager/latest/apireference/Welcome.html) functionality:

* [GetSecretValue](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html)
* [ListSecrets](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_ListSecrets.html)

Other functionality is not present yet, but could be added if the need arises (or someone contributes it).

TODO
----

See the TODO file.

BUGS
----

See the BUGS file.

Author
------

Jeremy Bingham (<jeremy.bingham@apexlearning.com>)

Copyright
---------

Copyright 2018, Apex Learning, Inc.

AWS Secrets Manager is copyright Amazon Web Services, Inc. (or maybe its affiliates, or even its parent).

License
-------

fake-secretsmanager is licensed under the terms of the Apache 2.0 License. See the LICENSE file for details.


CHANGELOG
0.2.0
-----
* Add ListSecrets to fake-secretsmanager's functionality.

0.1.0
-----
* Initial and early release.


get-secret_test.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"net/http"
	"testing"
)

const (
	goodkey = "good/secret/key"
	badkey  = "bad/secret/key"
	goodval = "i'm a secret, shhh"
)

func init() {
	secretMap = make(map[string]string)
	secretMap[goodkey] = goodval
	secretMap["foo/bar/bug"] = "sosecret"
	secretMap["fizz/bizz/buzz/blug"] = "1234568899090903130443"
}

func TestGetSecret(t *testing.T) {
	fetching := map[string]interface{}{"SecretId": goodkey}
	s, err := getSecret(fetching)
	if s == nil {
		t.Errorf("secret data was nil!")
	}
	if err != nil {
		t.Errorf("getSecret error is: %+v", err)
	}

	if s != nil && s.SecretString != goodval {
		t.Errorf("SecretString should have been '%s', but was '%s' instead.", goodval, s.SecretString)
	}
}

func TestGetNotExistSecret(t *testing.T) {
	fetching := map[string]interface{}{"SecretId": badkey}
	s, err := getSecret(fetching)
	if s != nil {
		t.Errorf("secret data was not nil when it should have been. Struct contains: %+v", s)
	}
	if err != nil && err.Status() != http.StatusBadRequest {
		t.Errorf("err status should have been %d, but was %d", http.StatusBadRequest, err.Status())
	}
}

func TestEmptySecretId(t *testing.T) {
	fetching := make(map[string]interface{})
	_, err := getSecret(fetching)
	if err == nil {
		t.Errorf("trying to fetch a secret with no supplied SecretId failed without an error")
	} else if err != nil && err.Error() != "no SecretId found in request!" {
		t.Errorf("Trying to fetch a secret without a SecretId failed, but for the wrong reason. Error was: '%s'.", err.Error())
	}

}


list-secrets_test.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"fmt"
	"testing"
)

func TestListSecrets(t *testing.T) {
	// empty - not using any of the values that could be passed in.
	r := make(map[string]interface{}) 
	list, _ := listSecrets(r)

	expectedLen := len(secretMap)
	if len(list.Items) != expectedLen {
		t.Errorf("There should have been %d items in the secrets list, but there were %d instead.", expectedLen, len(list.Items))
	}

	if list.NextToken != nil {
		t.Errorf("Strange - list.NextToken is not nil, but it should always be currently: %v", *list.NextToken)
	}

	i := 0
	for k, _ := range secretMap {
		if k != *list.Items[i].Name {
			t.Errorf("Item %d's name should have been '%s', but was '%s'.", i, k, *list.Items[i].Name)
		}
		testArn := makeArn(k)
		if testArn != *list.Items[i].Arn {
			t.Errorf("Item %d ARN should be '%s', but is '%s'", i, testArn, *list.Items[i].Arn)
		}
		expectedDesc := fmt.Sprintf("A pretend secret, id '%s'", k)
		if expectedDesc != *list.Items[i].Description {
			t.Errorf("Expected item %d description:\n%s\n\nActual Description:\n%s\n\n", i, expectedDesc, *list.Items[i].Description)
		}
		i++
	}
}


fake-secretsmanager_test.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"net/http"
	"testing"
)

func TestExceptionType(t *testing.T) {
	e := exceptionType(http.StatusBadRequest)
	if e != resourceNotFound {
		t.Errorf("Wrong exception type: should have been '%s', got '%s'", resourceNotFound, e)
	}
	e = exceptionType(http.StatusInternalServerError)
	if e != internalServiceErr {
		t.Errorf("Wrong exception type: should have been '%s', got '%s'", internalServiceErr, e)
	}
	e = exceptionType(http.StatusOK)
	if e != unknownException {
		t.Errorf("Wrong exception type: should have been '%s', got '%s'", unknownException, e)
	}
}

func TestMakeVersionId(t *testing.T) {
	secretId := "go-test/foo/bar/baz/bug/bleh"
	expectedId := "SECRET-6261722f62617a2f6275672f626c6568"
	vId := makeVersionId(secretId)

	if vId != expectedId {
		t.Errorf("wrong VersionId returned: expected '%s', got '%s'.", expectedId, vId)
	}
	if len(vId) > 64 || len(vId) < 32 {
		t.Errorf("incorrect VersionId length: it should be less than 64 characters and more than 32, but was %d.", len(vId))
	}
	short := "short"
	shortId := makeVersionId(short)
	if len(shortId) > 64 || len(shortId) < 32 {
		t.Errorf("incorrect VersionId length for short secret key: it should be less than 64 characters and more than 32, but was %d.", len(shortId))
	}

	long := "longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong"
	longId := makeVersionId(long)
	if len(longId) > 64 || len(longId) < 32 {
		t.Errorf("incorrect VersionId length for long secret key: it should be less than 64 characters and more than 32, but was %d.", len(longId))
	}
}


util.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"encoding/hex"
	"encoding/json"
	"fmt"
	"github.com/apexlearning/fake-secretsmanager/internal/smerror"
	"io"
	"log"
	"net/http"
)

/* Various utility functions that may be useful in multiple places. */

const arnBase = "arn:aws:secretsmanager:%s:%012d:secret:%s"

func makeArn(secretId string) string {
	return fmt.Sprintf(arnBase, region, accountId, secretId)
}

func makeVersionId(secretId string) string {
	b := []byte(secretId)
	idLen := 16
	if len(b) < 16 {
		k := make([]byte, 16-len(b))
		b = append(k, b...)
	}
	str := hex.EncodeToString(b)
	vId := fmt.Sprintf("SECRET-%s", str[len(str)-(idLen*2):])
	return vId
}

func parseJSON(data io.ReadCloser) (map[string]interface{}, smerror.Error) {
	reqData := make(map[string]interface{})
	dec := json.NewDecoder(data)
	if err := dec.Decode(&reqData); err != nil {
		smerr := smerror.CastErr(err)
		// 500 I *think* is probably most appropriate here, but 400
		// might be more appropriate.
		smerr.SetStatus(http.StatusInternalServerError)
		return nil, smerr
	}
	return reqData, nil
}

func jsonErrorReport(w http.ResponseWriter, r *http.Request, errorStr string, status int) {
	log.Println(errorStr)

	jsonError := make(map[string]string)
	jsonError["Message"] = errorStr
	jsonError["__type"] = exceptionType(status)

	w.WriteHeader(status)
	enc := json.NewEncoder(w)
	if err := enc.Encode(&jsonError); err != nil {
		log.Println(err)
	}
	return
}

func exceptionType(status int) string {
	switch status {
	case http.StatusBadRequest:
		return resourceNotFound
	case http.StatusInternalServerError:
		return internalServiceErr
	default:
		return unknownException
	}
}


config.go
package main

import (
	"fmt"
	"github.com/jessevdk/go-flags"
	"os"
	"runtime"
)

const (
	version     = "0.2.0"
	defaultAddr = ":7887"
)

var GitHash = "unknown"

type options struct {
	Version     bool   `short:"v" long:"version" description:"Print version info."`
	Addr        string `short:"a" long:"addr" description:"IP address to listen on. Default: ':7887'." env:"FAKESM_ADDR"`
	SecretsJson string `short:"f" long:"secrets-json" description:"Path to JSON file containing the secrets in a hash. The JSON hash key names are the secret names. If the secret is itself JSON, it needs to be escaped and stuffed in there as a normal string." env:"FAKESM_SECRETS_JSON"`
}

func parseOptions() (*options, error) {
	var opts = &options{}
	parser := flags.NewParser(opts, flags.Default)
	parser.ShortDescription = fmt.Sprintf("A stand-in for AWS Secrets Manager for local testing that doesn't require access to the real Secrets Manager - version %s", version)
	parser.NamespaceDelimiter = "-"

	_, err := parser.Parse()
	if err != nil {
		if err.(*flags.Error).Type == flags.ErrHelp {
			os.Exit(0)
		} else {
			return nil, err
		}
	}
	if opts.Version {
		fmt.Printf("fake-secretsmanager %s (git hash: %s) built with %s.\n", version, GitHash, runtime.Version())
		os.Exit(0)
	}
	if opts.Addr == "" {
		opts.Addr = defaultAddr
	}
	if opts.SecretsJson == "" {
		return nil, fmt.Errorf("No secrets file provided!")
	}

	return opts, nil
}


get-secret.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"github.com/apexlearning/fake-secretsmanager/internal/smerror"
	"net/http"
)

type secretVersion struct {
	Arn           string   `json:"ARN"`
	CreatedDate   int64    `json:"CreatedDate"`
	Name          string   `json:"Name"`
	SecretBinary  []byte   `json:"SecretBinary"`
	SecretString  string   `json:"SecretString"`
	VersionId     string   `json:"VersionId"`
	VersionStages []string `json:"VersionStages"`
}

func getSecret(data map[string]interface{}) (*secretVersion, smerror.Error) {
	var secretId string
	if sid, ok := data["SecretId"]; ok {
		secretId = sid.(string)
	} else {
		smerr := smerror.Errorf("no SecretId found in request!")
		smerr.SetStatus(http.StatusInternalServerError)
		return nil, smerr
	}

	s := new(secretVersion)
	s.Arn = makeArn(secretId)
	s.CreatedDate = setTimestamp
	s.Name = secretId
	if scr, ok := secretMap[secretId]; !ok {
		return nil, smerror.Errorf("secret '%s' not found", secretId)
	} else {
		s.SecretString = scr
	}
	s.VersionId = makeVersionId(secretId)
	s.VersionStages = []string{"AWSCURRENT"}

	return s, nil
}


list-secrets.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"fmt"
	"github.com/apexlearning/fake-secretsmanager/internal/smerror"
)

/*
 * yaaaaaaaaaaaay pointers! Needed here to make the JSON response behave
 * correctly, though.
 */

type secretListItem struct {
	Arn                    *string             `json:"ARN"`
	DeletedDate            *int64              `json:"DeletedDate"`
	Description            *string             `json:"Description"`
	KmsKeyId               *string             `json:"KmsKeyId"`
	LastAccessedDate       *int64              `json:"LastAccessedDate"`
	LastChangedDate        *int64              `json:"LastChangedDate"`
	LastRotatedDate        *int64              `json:"LastRotatedDate"`
	Name                   *string             `json:"Name"`
	RotationEnabled        *bool               `json:"RotationEnabled"`
	RotationLambdaArn      *string             `json:"RotationLambdaARN"`
	RotationRules          map[string]int      `json:"RotationRules"`
	SecretVersionsToStages map[string][]string `json:"SecretVersionsToStages"`
	Tags                   map[string]string   `json:"Tags"`
}

type secretList struct {
	NextToken *string
	Items     []*secretListItem `json:"SecretList"`
}

func listSecrets(data map[string]interface{}) (*secretList, smerror.Error) {
	// for now, don't worry about paging or limits. That can be implemented
	// later if the need arises.
	_ = data

	listOfSecrets := new(secretList)
	listOfSecrets.Items = make([]*secretListItem, len(secretMap))

	stages := []string{"AWSCURRENT"}

	i := 0
	for k, _ := range secretMap {
		name := k
		arn := makeArn(name)
		desc := fmt.Sprintf("A pretend secret, id '%s'", name)
		vId := makeVersionId(name)

		item := new(secretListItem)
		item.Arn = &arn
		item.Name = &name
		item.LastChangedDate = &setTimestamp
		item.Description = &desc
		item.SecretVersionsToStages = map[string][]string{vId: stages}

		listOfSecrets.Items[i] = item
		i++
	}
	return listOfSecrets, nil
}


fake-secretsmanager.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"encoding/json"
	"fmt"
	"github.com/pborman/uuid"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	region               = "us-west-2"
	contentType          = "application/x-amz-json-1.1"
	getSecretTarget      = "secretsmanager.GetSecretValue"
	describeSecretTarget = "secretsmanager.DescribeSecret"
	listSecretsTarget    = "secretsmanager.ListSecrets"
	resourceNotFound     = "ResourceNotFoundException"
	internalServiceErr   = "InternalServiceError"
	unknownException     = "UnknownException"
)

const accountId = 123456789012

var secretMap map[string]string
var setTimestamp int64

func init() {
	secretMap = make(map[string]string)
	setTimestamp = time.Now().Unix()
}

func main() {
	opts, err := parseOptions()
	if err != nil {
		log.Fatal(err)
	}

	secretDataFile, err := os.Open(opts.SecretsJson)
	if err != nil {
		log.Fatal(err)
	}
	defer secretDataFile.Close()

	b, _ := ioutil.ReadAll(secretDataFile)
	err = json.Unmarshal(b, &secretMap)
	if err != nil {
		log.Fatal(err)
	}
	keys := make([]string, len(secretMap))
	i := 0
	for k, _ := range secretMap {
		keys[i] = k
		i++
	}
	log.Printf("Loading secrets with keys %s...", strings.Join(keys, ", "))
	http.HandleFunc("/", rootHandler)
	err = http.ListenAndServe(opts.Addr, nil)
	if err != nil {
		log.Println(err)
	}
}

func rootHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", contentType)
	reqId := uuid.New()
	w.Header().Set("X-Amz-RequestId", reqId)

	req, err := parseJSON(r.Body)
	if err != nil {
		jsonErrorReport(w, r, err.Error(), err.Status())
		return
	}

	switch r.Header.Get("x-amz-target") {
	case getSecretTarget:
		val, err := getSecret(req)
		if err != nil {
			jsonErrorReport(w, r, err.Error(), err.Status())
			return
		}
		enc := json.NewEncoder(w)
		if perr := enc.Encode(&val); perr != nil {
			log.Println(perr)
		}
	case describeSecretTarget:
		jsonErrorReport(w, r, "Target secretsmanager.DescribeSecret hasn't been implemented yet.", http.StatusInternalServerError)
	case listSecretsTarget:
		list, err := listSecrets(req)
		if err != nil {
			jsonErrorReport(w, r, err.Error(), err.Status())
			return
		}
		enc := json.NewEncoder(w)
		if perr := enc.Encode(&list); perr != nil {
			log.Println(perr)
		}
	default:
		t := r.Header.Get("x-amz-target")
		jsonErrorReport(w, r, fmt.Sprintf("Unimplemented target %s", t), http.StatusInternalServerError)
	}
	return
}


internal/smerror/smerror.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package smerror

import (
	"fmt"
	"net/http"
)

type smerror struct {
	msg    string
	status int
}

type Error interface {
	String() string
	Error() string
	Status() int
	SetStatus(int)
}

func New(text string) Error {
	return &smerror{msg: text,
		status: http.StatusBadRequest,
	}
}

func Errorf(format string, a ...interface{}) Error {
	return New(fmt.Sprintf(format, a...))
}

func CastErr(err error) Error {
	return Errorf(err.Error())
}

func (e *smerror) Error() string {
	return e.msg
}

func (e *smerror) String() string {
	return e.msg
}

func (e *smerror) SetStatus(s int) {
	e.status = s
}

func (e *smerror) Status() int {
	return e.status
}


doc.go
/*
 * Copyright (c) 2018, Apex Learning, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
'fake-secretsmanager' is a stand-in for the full AWS Secrets Manager (https://aws.amazon.com/secrets-manager/) for testing, local development, kitchen runs, and other such things where using the real deal is neither needed nor desirable.

It should go without saying that this absolutely should not be used for production (or probably even staging) - it does not use SSL to encrypt traffic, and its "secrets" storage is in a plain JSON file.

Installation

To build 'fake-secretsmanager' from source, assuming you have Go installed and your Go development environment set up properly, run:

	$ go get -u github.com/apexlearning/fake-secretsmanager

'fake-secretsmanager' has been built using golang 1.10. It may build with earlier or later versions of the compiler, but this has not been tested.

A 'Dockerfile' is also provided, if running 'fake-secretsmanager' inside docker is more convenient. Build it with 'docker build -t your-handy-name/fake-secretsmanager .', and run it with:

	$ docker run -p 7887:7887 -v /path/to/your/secrets.json:/opt/fake-secretsmanager/data/secrets.json --name fakesm -d your-handy-name/fake-secretsmanager

Usage

From the output of fake-secretsmanager's help argument:

	Usage:
	  fake-secretsmanager [OPTIONS]

	Application Options:
	  -v, --version       Print version info.
	  -a, --addr=         IP address to listen on. Default: ':7887'. [$FAKESM_ADDR]
	  -f, --secrets-json= Path to JSON file containing the secrets in a hash. The
			      JSON hash key names are the secret names. If the secret
			      is itself JSON, it needs to be escaped and stuffed in
			      there as a normal string. [$FAKESM_SECRETS_JSON]

	Help Options:
	  -h, --help          Show this help message

The 'secrets.json' file provides an example of how to format the JSON file that stores the secrets. Each secret must be a string. If it is not a string, it must be properly quoted and escaped as per the example escaped JSON in that file.

To use 'fake-secretsmanager', supply a custom endpoint to the AWS cli or in your code's secretsmanager client constructor.

CLI:

	$ aws secretsmanager get-secret-value --endpoint-url http://localhost:7887 --secret-id foo/json/escaped/sssssh

	$ aws secretsmanager list-secrets --endpoint http://localhost:7887

Ruby:

	secretsmanager = Aws::SecretsManager::Client.new(
	  region: my_region,
	  endpoint: 'http://localhost:7887',
	  # .....
	)

Supported Functionality

Currently, 'fake-secretsmanager' supports the following AWS Secrets Manager API (https://docs.aws.amazon.com/secretsmanager/latest/apireference/Welcome.html) functionality:

* GetSecretValue (https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html)
* ListSecrets (https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_ListSecrets.html)

Other functionality is not present yet, but could be added if the need arises (or someone contributes it).

TODO

See the TODO file.

BUGS

See the BUGS file.

Author

Jeremy Bingham (<jeremy.bingham@apexlearning.com>)

Copyright

Copyright 2018, Apex Learning, Inc.

AWS Secrets Manager is copyright Amazon Web Services, Inc. (or maybe its affiliates, or even its parent).

License

fake-secretsmanager is licensed under the terms of the Apache 2.0 License. See the LICENSE file for details.
*/
package main
